spec/workflows/estimate_spec.rb

```diff
@@ -0,0 +1,178 @@
+          f.puts({
+            "items_matched" => 1,
+            "items_ic" => 0,
+            "items_pd" => 1,
+            "h_share" => 0
+          }.to_json)
+        end
+      end
+
+      it "sums input json files to produce output in the expected format" do
+        described_class.new(working_directory: ENV["TEST_TMP"], ocn_file: ocn_file).run
+
+        estimate_file = File.join(Settings.estimates_path, "ocnfile-estimate-#{Date.today}.txt")
+
+        expect(File.read(estimate_file)).to eq <<~EOT
+          Total Estimated IC Cost: $5.00
+          In all, we received 4 distinct OCLC numbers.
+          Of those distinct OCLC numbers, 2 (50.0%) match items in
+          HathiTrust, corresponding to 2 HathiTrust items.
+          Of those items, 1 (50.0%) are in the public domain,
+          1 (50.0%) are in copyright.
+        EOT
+      end
+    end
+  end
+
+  describe Workflows::Estimate::Analyzer do
+    let(:test_records) { File.join(ENV["TEST_TMP"], record_filename) }
+    let(:output_json) do
+      JSON.parse(File.read(test_records + ".estimate.json"))
+    end
+
+    before(:each) do
+      FileUtils.cp(fixture(record_filename), ENV["TEST_TMP"])
+    end
+
+    context "with two items, one pd and one ic" do
+      let(:record_filename) { "records_for_estimate.ndj" }
+
+      it "outputs number of items matched" do
+        described_class.new(test_records).run
+        expect(output_json["items_matched"]).to eq(2)
+      end
+
+      it "outputs number of pd items matched" do
+        described_class.new(test_records).run
+        expect(output_json["items_pd"]).to eq(1)
+      end
+
+      it "outputs number of ic items matched" do
+        described_class.new(test_records).run
+        expect(output_json["items_ic"]).to eq(1)
+      end
+
+      it "outputs h_share_total" do
+        described_class.new(test_records).run
+        # the contributor gets the other half of the IC item
+        expect(output_json["h_share"]).to eq(0.5)
+      end
+    end
+
+    context "one with icus item" do
+      let(:record_filename) { "icus_item.ndj" }
+
+      it "outputs icus items as pd" do
+        FileUtils.cp(fixture("icus_item.ndj"), ENV["TEST_TMP"])
+        described_class.new(File.join(ENV["TEST_TMP"], "icus_item.ndj")).run
+        expect(output_json["items_pd"]).to eq(1)
+      end
+    end
+  end
+
+  describe Workflows::Estimate::DataSource do
+    let(:ocn_file) { File.join(ENV["TEST_TMP"], "ocnfile") }
+    let(:output) { File.join(ENV["TEST_TMP"], "records.out") }
+    let(:ocn_count) { File.join(ENV["TEST_TMP"], "ocn_count.estimate.json") }
+
+    let(:data_source) { described_class.new(ocn_file: ocn_file) }
+
+    describe "#dump_records" do
+      before(:each) { File.write(ocn_file, ocns.join("\n")) }
+
+      context "with four ocns and two items" do
+        let(:ocns) { [1, 2, ht_allow.ocns, ht_deny.ocns].flatten }
+        let(:ht_allow) { build(:ht_item, rights: "pd") }
```

Comment:
The `counts input & matching ocns` test below will fail if either `ht_allow` and/or `ht_deny` get assigned ocn 1/2 from the factory, or if they get the same ocn.
I would specify ocns 3 & 4 for `ht_allow` & `ht_deny` to remove this brittleness.

Reply:
I will go ahead and specify 3/4.

One thing I saw recently with `factory_bot` that I think would help significantly with some of these potentially brittle tests around ocns is sequences: https://thoughtbot.github.io/factory_bot/sequences/inline-sequences.html

Reply:
https://hathitrust.atlassian.net/browse/ETT-118, but I think the sequences would be good to try using the next time we find ourselves writing a test like this.